src/mono/mono/sgen/sgen-gc.c

```diff
@@ -4083,4 +4084,96 @@ sgen_check_canary_for_object (gpointer addr)
 	}
 }
 
+
+#define MEM_PRESSURE_COUNT  4
+#define MAX_MEMORYPRESSURE_RATIO  10  
+guint64 memory_pressure_gc_count = 0;
+guint64 memory_pressure_iteration = 0;
+guint64 memory_pressure_adds[MEM_PRESSURE_COUNT] = {0, 0, 0, 0}; 
+guint64 memory_pressure_removes[MEM_PRESSURE_COUNT] = {0, 0, 0, 0};   // history of memory pressure removals
+const unsigned min_memorypressure_budget = 4 * 1024 * 1024;        // 4 MB
+
+// Resets pressure accounting after a gen2 GC has occurred.
+static void check_pressure_counts ()
+{
+    if (memory_pressure_gc_count != sgen_gc_collection_count(GENERATION_OLD))
+    {
+        memory_pressure_gc_count = sgen_gc_collection_count(GENERATION_OLD);
+        memory_pressure_iteration++;
+
+        guint32 p = memory_pressure_iteration % MEM_PRESSURE_COUNT;
+
+        memory_pressure_adds[p] = 0;   // new pressure will be accumulated here
+        memory_pressure_removes[p] = 0;
+    }
+}
+
+void sgen_remove_memory_pressure (guint64 bytes_allocated)
+{
+    check_pressure_counts();
+
+    guint32 p = memory_pressure_iteration % MEM_PRESSURE_COUNT;
+
+    mono_atomic_fetch_add_i64((gint64*)&memory_pressure_removes[p], bytes_allocated);
+}
+
+void sgen_add_memory_pressure (guint64 bytes_allocated)
+{
+
+    check_pressure_counts ();
+
+    guint p = memory_pressure_iteration % MEM_PRESSURE_COUNT;
+
+    guint64 new_mem_value = mono_atomic_fetch_add_i64 ((gint64*)&memory_pressure_adds[p], bytes_allocated);
+    
+    // AddMemoryPressure contains unrolled loops which depend on MEM_PRESSURE_COUNT
+    g_assert (MEM_PRESSURE_COUNT == 4);
+
+    guint64 add = memory_pressure_adds[0] + memory_pressure_adds[1] + memory_pressure_adds[2] + memory_pressure_adds[3] - memory_pressure_adds[p];
+    guint64 rem = memory_pressure_removes[0] + memory_pressure_removes[1] + memory_pressure_removes[2] + memory_pressure_removes[3] - memory_pressure_removes[p];
+
+    if (new_mem_value >= min_memorypressure_budget)
```

Comment:
The rest of the code here seems to do two things

1. Evaluates a heuristic to decide if it should run a collection
2. Runs a collection.

Can we make those into separate `static` functions